File: BE/communication_service/src/main/java/com/ds/communication_service/infrastructure/kafka/AuditEventPublisher.java
package com.ds.communication_service.infrastructure.kafka;

import com.ds.communication_service.infrastructure.kafka.dto.AuditEventDto;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.kafka.support.SendResult;
import org.springframework.stereotype.Component;

import java.time.LocalDateTime;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;

/**
 * Publisher for audit events to Kafka
 * Handles retry logic and dead letter queue
 */
@Slf4j
@Component
@RequiredArgsConstructor
public class AuditEventPublisher {

    private final KafkaTemplate<String, Object> kafkaTemplate;

    /**
     * Publish audit event to Kafka
     * Uses async send with callback for error handling
     */
    public void publishAuditEvent(AuditEventDto event) {
        try {
            // Generate event ID if not provided
            if (event.getEventId() == null || event.getEventId().isBlank()) {
                event.setEventId(UUID.randomUUID().toString());
            }
            
            // Set timestamp if not provided
            if (event.getTimestamp() == null) {
                event.setTimestamp(LocalDateTime.now());
            }
            
            // Set source service
            if (event.getSourceService() == null || event.getSourceService().isBlank()) {
                event.setSourceService("communication-service");
            }
            
            // Use resourceId or userId as key for partitioning
            String key = event.getResourceId() != null 
                ? event.getResourceId() 
                : (event.getUserId() != null ? event.getUserId() : event.getEventId());
            
            // Send to Kafka asynchronously
            CompletableFuture<SendResult<String, Object>> future = kafkaTemplate.send(
                KafkaConfig.TOPIC_AUDIT_EVENTS,
                key,
                event
            );
            
            // Add callback for error handling
            future.whenComplete((result, ex) -> {
                if (ex == null) {
                    log.debug("✅ Audit event published successfully: eventId={}, operation={}, resourceType={}", 
                        event.getEventId(), event.getOperationType(), event.getResourceType());
                } else {
                    log.error("❌ Failed to publish audit event: eventId={}, operation={}, resourceType={}", 
                        event.getEventId(), event.getOperationType(), event.getResourceType(), ex);
                    
                    // Send to Dead Letter Queue for manual review
                    sendToDeadLetterQueue(event, ex);
                }
            });
            
        } catch (Exception e) {
            log.error("❌ Error publishing audit event: {}", e.getMessage(), e);
            // Send to Dead Letter Queue
            sendToDeadLetterQueue(event, e);
        }
    }

    /**
     * Send failed event to Dead Letter Queue
     */
    private void sendToDeadLetterQueue(AuditEventDto event, Throwable error) {
        try {
            // Mark as failed
            event.setStatus(AuditEventDto.Status.FAILED);
            event.setErrorMessage(error.getMessage());
            
            // Truncate stack trace if too long (max 5000 chars)
            if (error.getStackTrace() != null) {
                StringBuilder stackTrace = new StringBuilder();
                for (StackTraceElement element : error.getStackTrace()) {
                    stackTrace.append(element.toString()).append("\n");
                    if (stackTrace.length() > 5000) {
                        stackTrace.append("... (truncated)");
                        break;
                    }
                }
                event.setErrorStackTrace(stackTrace.toString());
            }
            
            // Send to DLQ
            kafkaTemplate.send(KafkaConfig.TOPIC_AUDIT_EVENTS_DLQ, event.getEventId(), event);
            log.debug("[communication-service] [AuditEventPublisher.sendToDLQ] Audit event sent to DLQ: eventId={}", event.getEventId());
            
        } catch (Exception e) {
            log.error("[communication-service] [AuditEventPublisher.sendToDLQ] CRITICAL: Failed to send audit event to DLQ", e);
            // At this point, we can only log to application logs
            // Consider implementing a fallback mechanism (e.g., file-based logging)
        }
    }

    /**
     * Helper method to create audit event from operation details
     */
    public void logOperation(
            AuditEventDto.OperationType operationType,
            String httpMethod,
            String endpoint,
            String resourceType,
            String resourceId,
            String userId,
            String userRoles,
            Integer responseStatus,
            AuditEventDto.Status status,
            Long durationMs,
            String requestId,
            String clientIp,
            String userAgent,
            Map<String, Object> requestPayload,
            String errorMessage) {
        
        AuditEventDto event = AuditEventDto.builder()
            .operationType(operationType)
            .httpMethod(httpMethod)
            .endpoint(endpoint)
            .resourceType(resourceType)
            .resourceId(resourceId)
            .userId(userId)
            .userRoles(userRoles)
            .responseStatus(responseStatus)
            .status(status)
            .durationMs(durationMs)
            .requestId(requestId)
            .clientIp(clientIp)
            .userAgent(userAgent)
            .requestPayload(requestPayload)
            .errorMessage(errorMessage)
            .build();
        
        publishAuditEvent(event);
    }
}

File: BE/zone_service/src/common/kafka/audit-event-publisher.ts
/**
 * Audit Event Publisher
 * Publishes audit events to Kafka for all CRUD operations
 */

import { kafkaService } from './kafka.service';
import { logger } from '../logger/logger.service';
import { randomUUID } from 'crypto';

const TOPIC_AUDIT_EVENTS = 'audit-events';
const TOPIC_AUDIT_EVENTS_DLQ = 'audit-events-dlq';

export enum OperationType {
  CREATE = 'CREATE',
  UPDATE = 'UPDATE',
  DELETE = 'DELETE',
  READ = 'READ',
  CONNECT = 'CONNECT',
  DISCONNECT = 'DISCONNECT',
  MESSAGE = 'MESSAGE',
}

export enum Status {
  SUCCESS = 'SUCCESS',
  FAILED = 'FAILED',
  PENDING = 'PENDING',
}

export interface AuditEventDto {
  eventId?: string;
  timestamp?: string; // ISO 8601 format
  operationType: OperationType;
  sourceService?: string;
  userId?: string;
  userRoles?: string;
  httpMethod?: string;
  endpoint?: string;
  resourceType?: string;
  resourceId?: string;
  requestPayload?: Record<string, any>;
  responseStatus?: number;
  status?: Status;
  errorMessage?: string;
  errorStackTrace?: string;
  requestId?: string;
  metadata?: Record<string, any>;
  durationMs?: number;
  clientIp?: string;
  userAgent?: string;
}

class AuditEventPublisher {
  private static instance: AuditEventPublisher | null = null;

  private constructor() {}

  /**
   * Get singleton instance
   */
  public static getInstance(): AuditEventPublisher {
    if (!AuditEventPublisher.instance) {
      AuditEventPublisher.instance = new AuditEventPublisher();
    }
    return AuditEventPublisher.instance;
  }

  /**
   * Publish audit event to Kafka
   * Uses async send with error handling
   */
  public async publishAuditEvent(event: AuditEventDto): Promise<void> {
    try {
      // Generate event ID if not provided
      if (!event.eventId) {
        event.eventId = randomUUID();
      }

      // Set timestamp if not provided
      if (!event.timestamp) {
        event.timestamp = new Date().toISOString();
      }

      // Set source service
      if (!event.sourceService) {
        event.sourceService = 'zone-service';
      }

      // Use resourceId or userId as key for partitioning
      const key = event.resourceId || event.userId || event.eventId || 'unknown';

      // Skip if Kafka is not connected
      if (!kafkaService.getConnectionStatus()) {
        logger.warn('[AuditEventPublisher] Kafka not connected, audit event will be skipped', {
          eventId: event.eventId,
          operation: event.operationType,
        });
        return;
      }

      // Send to Kafka asynchronously (non-blocking)
      await kafkaService.sendMessage(TOPIC_AUDIT_EVENTS, event, key);

      logger.debug('[AuditEventPublisher] Audit event published successfully', {
        eventId: event.eventId,
        operation: event.operationType,
        resourceType: event.resourceType,
      });
    } catch (error: any) {
      logger.error('[AuditEventPublisher] Failed to publish audit event', {
        eventId: event.eventId,
        operation: event.operationType,
        error: error.message,
      });

      // Send to Dead Letter Queue for manual review
      await this.sendToDeadLetterQueue(event, error);
    }
  }

  /**
   * Send failed event to Dead Letter Queue
   */
  private async sendToDeadLetterQueue(event: AuditEventDto, error: Error): Promise<void> {
    try {
      // Mark as failed
      event.status = Status.FAILED;
      event.errorMessage = error.message;

      // Truncate stack trace if too long (max 5000 chars)
      if (error.stack) {
        const stackTrace = error.stack.length > 5000
          ? error.stack.substring(0, 5000) + '... (truncated)'
          : error.stack;
        event.errorStackTrace = stackTrace;
      }

      // Skip if Kafka is not connected
      if (!kafkaService.getConnectionStatus()) {
        logger.error('[AuditEventPublisher] Kafka not connected, cannot send to DLQ', {
          eventId: event.eventId,
        });
        return;
      }

      // Send to DLQ
      await kafkaService.sendMessage(TOPIC_AUDIT_EVENTS_DLQ, event, event.eventId || 'unknown');
      logger.debug('[AuditEventPublisher] Audit event sent to DLQ', {
        eventId: event.eventId,
      });
    } catch (dlqError: any) {
      logger.error('[AuditEventPublisher] CRITICAL: Failed to send audit event to DLQ', {
        eventId: event.eventId,
        error: dlqError.message,
      });
      // At this point, we can only log to application logs
    }
  }

  /**
   * Helper method to create audit event from operation details
   */
  public async logOperation(
    operationType: OperationType,
    httpMethod?: string,
    endpoint?: string,
    resourceType?: string,
    resourceId?: string,
    userId?: string,
    userRoles?: string,
    responseStatus?: number,
    status: Status = Status.SUCCESS,
    durationMs?: number,
    requestId?: string,
    clientIp?: string,
    userAgent?: string,
    requestPayload?: Record<string, any>,
    errorMessage?: string
  ): Promise<void> {
    const event: AuditEventDto = {
      operationType,
      status,
      ...(httpMethod !== undefined && { httpMethod }),
      ...(endpoint !== undefined && { endpoint }),
      ...(resourceType !== undefined && { resourceType }),
      ...(resourceId !== undefined && { resourceId }),
      ...(userId !== undefined && { userId }),
      ...(userRoles !== undefined && { userRoles }),
      ...(responseStatus !== undefined && { responseStatus }),
      ...(durationMs !== undefined && { durationMs }),
      ...(requestId !== undefined && { requestId }),
      ...(clientIp !== undefined && { clientIp }),
      ...(userAgent !== undefined && { userAgent }),
      ...(requestPayload !== undefined && { requestPayload }),
      ...(errorMessage !== undefined && { errorMessage }),
    };

    await this.publishAuditEvent(event);
  }
}

// Export singleton instance
export const auditEventPublisher = AuditEventPublisher.getInstance();

// Export class for advanced use cases
export { AuditEventPublisher };
